Aerospike agent and dashboards

mphanias opened the pull request:
### What does this PR do?
Modified the aerospike integration to include all the metrics exposed by aerospike-pormetheus-exporter and multiple dashboard consuming the exported metrics. 

### Motivation
Expanding Metric Collection and Dashboards, we want to ensure all Aerospike metrics are available in datadog, so customers can have better visibility and provide actionable insights for key metrics, including performance, latency, and resource utilization.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Add the `qa/skip-qa` label if the PR doesn't need to be tested during QA.
- [ ] If you need to backport this PR to another branch, you can add the `backport/<branch-name>` label to the PR and it will automatically open a backport PR once this one is merged

iliakur reviewed (changes requested):
@mphanias thanks for taking the bull by the horns here!

I have a couple concerns with the PR as it is right now:

1. We should not drop support for metrics we already collect. This will cause significant pain to folks who use older versions.
2. I don't understand why tests were removed but no new ones were added.
3. For the dashboards, have you [had a look at these guidelines](https://datadoghq.dev/integrations-core/guidelines/dashboards/)?

mphanias commented:
Hi iliakur,

Thank you for your feedback, please find my response inline,

1. We should not drop support for metrics we already collect. This will cause significant pain to folks who use older versions.
    Response:
    With the new changes all the metrics exposed by Aeropike Prometheus Exporter will be sent without any filters irrespective of the Aerospike server version,
    Aerospike Prometheus Exporter is rewritten to be Aerospike DB server agnostic and will send all metrics/stats given by server to the caller in OpenMetrics format, 
    But surely we modified in existing Datadog Agent is, earlier metrics are read by DD Agent and sent as a different metric name which we changed.
   Why we have taken this approach is, Aerospike 7.0 version was released more than a year back and most customers are already on 7.0, and Aerospike 4.9, 5.x are end-of-life and soon Aerospike 6.0 also will be end-of-life.
  
2. I don't understand why tests were removed but no new ones were added.
    Response: 
    I agree removed test-cases, we did this because, with the proposed changes, DD Agent will now get all metrics from Aerospike Prometheus Exporter and metrics are sent with their names and values without any transformation within the agent (except prefixing each metric 'aerospike'), since we are not doing any transformation we felt test-cases for 5.x and 6.x are not required.

4. For the dashboards, have you [had a look at these guidelines](https://datadoghq.dev/integrations-core/guidelines/dashboards/)?
    Response:
    Thank you for pointing towards the guidelines, we looked at all the dashboards and ensured adhering to all the guidelines mentioned.

Regards

mphanias commented:
Hi iliakur,

We modified the existing overview dashboard to use the actual metric names to ensure existing customer are not impacted. if any custom dashboard are designed by customers using old naming convention would have impact, but current proposed changes will ensure all customers can get the full benefits of all the metrics exposed by Aerospike Prometheus Exporter.

Regards

mphanias commented:
Hi iliakur,

Could you please share your inputs or feedback on the comments I added, really appreciate your support.

Regards
Phani.

mphanias commented:
Hi iliakur,

I understand your point of supporting EOL customer, I will add back all the test-cases and then re-add my add changes. 

while testing these changes in my local machine, I am facing some issue, could you please point me to any guide.

appreciate your review feedback and support.

Regards
Phani.

error I am getting - datadog_checks.base.checks.base.aerospike:aerospike.py:91 The `aerospike` client is not installed: No module named 'aerospike'

mphanias commented:
Hi iliakur,

Correct the code and added back all the test-cases supporting older versions. please review and share your inputs. 

Regards
Phani.

iliakur commented:
@mphanias thanks! I will get back to you within the next 2 weeks. Under some load right now.

iliakur commented:
@mphanias I see no new changes. The commit history shows only merging master into your branch. Are you sure you pushed your changes that bring back the tests for old versions?

mphanias commented:
Hi iliakur,

Thank you for comments, I did a mistake and checking all my changes into the master branch and into my dev branch. I corrected all the files and submitted again, please review and share your comments.

Regards
Phani.

mphanias commented:
Hi iliakur,

Request your help in review. 

Regards
Phani.

mphanias commented:
Hi iliakur,

Hope you doing great, Request your help in review, we put back all the test-cases and added support for 7.x version also,  if you suggest any more changes I will incorporate them. 

Regards
Phani.

iliakur commented on aerospike/CHANGELOG.md:
```diff
@@ -2,6 +2,16 @@
 
 <!-- towncrier release notes start -->
 
+## 4.1.0 / 2025-02-25 / Agent 7.64.0
+
+***Added***:
+
+* New dashboards to show various important Aerospike metrics in multiple dashboards
+
+***Fixed***:
+
+* Allows all metrics from aerospike-prometheus-exporter to be available in datadog
+
```
```suggestion
```

iliakur commented on aerospike/changelog.d/19188.added:
```diff
@@ -0,0 +1,5 @@
+Sy/slurm user
+
```
```suggestion
```

iliakur commented on aerospike/changelog.d/19188.fixed:
```diff
@@ -0,0 +1,3 @@
+Sy/slurm user
+
```
```suggestion
```

iliakur commented on aerospike/tests/common.py:
```diff
@@ -4,7 +4,7 @@
 
 HERE = os.path.dirname(os.path.abspath(__file__))
-COMPOSE_FILE = os.path.join(HERE, 'docker', 'docker-compose.yaml')
+COMPOSE_FILE = os.path.join(HERE, 'docker', os.environ.get('DOCKER_COMPOSE_FILENAME'))
```
```suggestion
COMPOSE_FILE = os.path.join(HERE, 'docker', os.environ.['DOCKER_COMPOSE_FILENAME'])
```
Let's let this fail fast if the envvar isn't defined.

mphanias replied:
Accepted the review feedback

iliakur commented on aerospike/tests/docker/config/aerospike.conf:
```diff
@@ -0,0 +1,76 @@
+        group root
+        paxos-single-replica-limit 1 # Number of nodes where the replica count is automatically reduced to 1.
+        pidfile /var/run/aerospike/asd.pid
+#       service-threads 10 # cpu x 5 in 4.7 
+#       transaction-queues 10 # obsolete in 4.7 
+#       transaction-threads-per-queue 4 # obsolete in 4.7
```
```suggestion
```
Let's remove dead code

iliakur commented on aerospike/tests/docker/config/aerospike.conf:
```diff
@@ -0,0 +1,76 @@
+        memory-size 1G
+        default-ttl 30d # 5 days, use 0 to never expire/evict.
+        nsup-period 120
+        #       storage-engine memory
```
```suggestion
```

iliakur commented on aerospike/datadog_checks/aerospike/check.py:
```diff
@@ -1,22 +1,78 @@
+            self._fetch_build_info_from_metric()
+
+        # build-version is identified from node_up metrics, which is available from aerospike server 4.x
+        # dfault to 7 if no build version info is available, version number example: 7.1.0.2, 5.6.0.0
```
```suggestion
        # default to 7 if no build version info is available, version number example: 7.1.0.2, 5.6.0.0
```

iliakur commented on aerospike/hatch.toml:
```diff
@@ -4,15 +4,31 @@ check-types = false
 
 [envs.default.env-vars]
-AEROSPIKE_EXPORTER_VERSION = "1.5.0"
+AEROSPIKE_TOOLS_VERSION = "11.0.2"
```
Where is this used?

mphanias replied:
Thanks for pointing, fixed this now, this constant is used in the docker-compose file to fetch the required Aerospike-tools image

iliakur commented on aerospike/tests/test_unit_openmetricsv2.py:
```diff
@@ -18,19 +18,92 @@ def get_fixture_path(filename):
+
+    if version_parts[0] >= 7:
+        # these tests will include metrics from aerospike server version 7 and above
+        metrics_to_check = aggregator.metric_names
```
This makes the test useless. We're asserting that we collected the metrics that we collected. So if we miss a metric for whatever reason this test won't detect it.

Please add a hard-coded list of metrics that you expect to see submitted for a given payload.

mphanias replied:
@iliakur, I agree with your comment, as we did not have any fixed/static list in the metrics-map, with the new approach we are implementing using custom-transformer, we are now enhancing this test-case for v7.

mphanias replied:
We are adding 7x metrics as a static list in the commons.py for testing

iliakur commented on aerospike/tests/test_unit_openmetricsv2.py:
```diff
@@ -18,19 +18,92 @@ def get_fixture_path(filename):
-    for metric_name in EXPECTED_PROMETHEUS_METRICS + EXPECTED_PROMETHEUS_METRICS_5_6 + PROMETHEUS_XDR_METRICS:
+    version_parts = [int(p) for p in VERSION.split('.')]
+
+    if version_parts[0] >= 7:
```
Since we're providing the payloads as static text files (instead of connecting to an aerospike instance), we don't need to check the VERSION. You can just have 3 tests:
- `v7`
- `> v7`
- `< v7`

mphanias replied:
we have some metrics difference between 4.x and 5.x like "aerospike.node_stats.client_connections_opened.count" which are not available in 4.x causing our test-cases to fail, hence have tests for difference metrics. so >=7x , between 5.6 to 6.x and <5.6

iliakur commented on aerospike/tests/test_aerospike.py:
How come no tests for v7?

mphanias replied:
@iliakur, we added a test-case for v7 for metrics validation

iliakur commented on aerospike/datadog_checks/aerospike/metrics.py:
```diff
@@ -59,3 +59,16 @@
 }
+
+
+METRIC_MAP_V7 = {
```
Some issues with this setup:

- all metrics have the prefix `aerospike.aerospike_`. That's redundant, forces some users to type more, increases metric name length without adding signal.
- we break with the Datadog convention of using `.` to indicate namespaces. This would be less of an issue for a brand new integration, but for one where part of the metrics follow the convention and part don't that's bad UX.

mphanias replied:
@iliakur, 
Agree with you comment, previously we were not able to find a correct way to fix this name, now with the new custom-transformer based to convert the metric-names received from exporter, with the datadog standard naming convention like aerospike.namespace.master_objects ( i.e. we removed aerospike.aerospike_namespace_master_objects)

We modified the check.py using a custom transformer on exporter specific naming patterns, this will avoid a static metric-names map in metrics.py.

Reason we implemented this way was to send all the metrics from exporter/aerospike-server to so customer can use all of the metrics.

We will have consolidated and there were some changes in metric-names and type in latest exporter version, hence we did not attempt to merge 7.x metrics map with the previous metrics. 

please review and share your inputs

iliakur commented on aerospike/datadog_checks/aerospike/check.py:
```diff
@@ -1,22 +1,78 @@
+        """
+        shares details of metrics map depending on aerospike version 7 or below
+        """
+        metric_map_to_use = self._get_metrics_map()
```
I'm wondering if we need this logic to determine the metric map at runtime.

Is it the case that some metric in v7 has the same name as a metric in in <v7 but a different value or type?

If no, then we can have one big metric map that includes both v7 and non-v7 metrics and we let the contents of the payload determine which metrics get submitted.

iliakur reviewed (changes requested):
Thanks for keeping support for old metrics.

I still have serious concerns about the way we collect and name the metrics, that's a blocker.

Otherwise mostly small tweak suggestions.

iliakur commented:
@mphanias thanks for making the changes. There are some discussions outstanding that are blockers, please address them.

iliakur commented:
> @iliakur , I made the suggested changes, request your help in review and any further feedback.
> 
> Regards Phani.

Will get to this next week
